source/nodejs/adaptivecards/src/card-elements.ts

```diff
@@ -3545,6 +3545,242 @@ export class ChoiceSetInput extends Input {
+		element.style.width = "100%";
+
+		element.style.textAlign = "center";
+
+		this._toggleInputs = [];
+
+		let iconUnselected = this.iconUnselected ? this.iconUnselected : defaulticonUnselected;
+		let iconSelected = this.iconSelected ? this.iconSelected : defaulticonSelected;
+		let maxValue: number = this.maxValue ? this.maxValue : DEFAULT_RATING_SCALE;
+
+		let labels: Label[] = new Array(maxValue);
+		let labelElements: HTMLElement[] = new Array(maxValue);
+
+		for (let i = 0; i < maxValue; i++) {
+			let radioInput = document.createElement("input");
+			radioInput.id = Utils.generateUniqueId();
+			radioInput.type = "radio";
+			radioInput.style.margin = "0";
+			radioInput.style.display = "inline";
+			radioInput.style.verticalAlign = "middle";
+			radioInput.name = Utils.isNullOrEmpty(this.id) ? uniqueCategoryName : this.id;
+			radioInput.value = (i + 1).toString();
+			radioInput.style.flex = "0 0 auto";
+			radioInput.setAttribute("aria-label", "Rating " + (i + 1));
+			radioInput.style.display = "none";
+
+			radioInput.onchange = () => { this.valueChanged(); }
+
+			this._toggleInputs.push(radioInput);
+
+			labels[i] = new Label();
+			labels[i].setParent(this);
+			labels[i].forElementId = radioInput.id;
+			labels[i].hostConfig = this.hostConfig;
+			labels[i].text = "Rating " + (i + 1);
+			labels[i].useMarkdown = AdaptiveCard.useMarkdownInRadioButtonAndCheckbox;
+			labels[i].wrap = this.wrap;
+
+			labelElements[i] = labels[i].render();
+			labelElements[i].style.display = "block";
+			labelElements[i].style.flex = "1 1 auto";
+			labelElements[i].style.marginLeft = "6px";
+			labelElements[i].style.verticalAlign = "middle";
+			labelElements[i].style.flexGrow = "1";
+
+			labelElements[i].style.backgroundImage = "url('" + iconUnselected + "')";
+			labelElements[i].style.backgroundSize = "25px 25px";
+			labelElements[i].style.backgroundRepeat = "no-repeat";
+			labelElements[i].style.backgroundPositionX = "center";
+			labelElements[i].style.paddingTop = "25px";
+
+			let textFirstElementChild = <HTMLElement>labelElements[i].firstElementChild;
+			textFirstElementChild.style.width = "0%";
+			textFirstElementChild.style.removeProperty("text-overflow");
+			textFirstElementChild.style.overflow = "hidden";
+			textFirstElementChild.style.marginBottom = "-20px";
+
+			var ratingClicked: number = NOT_CLICKED;
+
+			// when hovering over an icon, replace that and all preceding icons with the iconSelected image
+			labelElements[i].onmouseover = function () {
+				for (let j = 0; j <= i; j++) {
+					labelElements[j].style.backgroundImage = "url('" + iconSelected + "')";
+				}
+				for (let j = i + 1; j < maxValue; j++) {
+					labelElements[j].style.backgroundImage = "url('" + iconUnselected + "')";
+				}
+			};
+
+			// when leaving an icon (i.e. we stop hovering), return to previous state
+			labelElements[i].onmouseleave = function () {
+				if (ratingClicked == NOT_CLICKED) {
+					for (let j = 0; j <= i; j++) {
```

Comment:
This logic of changing the look and feel of the icons should really be put in a separate function instead of being repeated several times. The function would take a value representing the value to reflect.

Also, the actual value of the rating control should be stored in a private member and updated as the radio inputs are clicked. Then the `onmouseleave` event can just call the helper method mentioned above with the current value of the input.